core/scheduler/handler/v1beta1/job_run.go

```diff
@@ -0,0 +1,190 @@
+		Envs:    input.Configs,
+		Files:   input.Files,
+		Secrets: input.Secrets,
+	}, nil
+}
+
+// JobRun currently gets the job runs from scheduler based on the criteria
+// TODO: later should collect the job runs from optimus
+func (h JobRunHandler) JobRun(ctx context.Context, req *pb.JobRunRequest) (*pb.JobRunResponse, error) {
+	projectName, err := tenant.ProjectNameFrom(req.GetProjectName())
+	if err != nil {
+		return nil, errors.GRPCErr(err, "unable to get job run for "+req.GetJobName())
+	}
+
+	jobName, err := scheduler.JobNameFrom(req.GetJobName())
+	if err != nil {
+		return nil, errors.GRPCErr(err, "unable to get job run for "+req.GetJobName())
+	}
+
+	criteria, err := buildCriteriaForJobRun(req)
+	if err != nil {
+		return nil, errors.GRPCErr(err, "unable to get job run for "+req.GetJobName())
+	}
+
+	var jobRuns []*scheduler.JobRunStatus
+	jobRuns, err = h.service.GetJobRuns(ctx, projectName, jobName, criteria) // TODO: return not found if not runs found
+	if err != nil {
+		if errors.IsErrorType(err, errors.ErrNotFound) {
+			return &pb.JobRunResponse{}, nil //todo: need strict review
+		}
+		return nil, errors.GRPCErr(err, "unable to get job run for "+req.GetJobName())
+	}
+
+	var runs []*pb.JobRun
+	for _, run := range jobRuns {
+		ts := timestamppb.New(run.ScheduledAt)
+		runs = append(runs, &pb.JobRun{
+			State:       run.State.String(),
+			ScheduledAt: ts,
+		})
+	}
+	return &pb.JobRunResponse{JobRuns: runs}, nil
+}
+
+func buildCriteriaForJobRun(req *pb.JobRunRequest) (*scheduler.JobRunsCriteria, error) {
+	if !req.GetStartDate().IsValid() && !req.GetEndDate().IsValid() {
+		return &scheduler.JobRunsCriteria{
+			Name:        req.GetJobName(),
+			OnlyLastRun: true,
+		}, nil
+	}
+	if !req.GetStartDate().IsValid() {
+		return nil, errors.InvalidArgument(scheduler.EntityJobRun, "empty start date is given")
+	}
+	if !req.GetEndDate().IsValid() {
+		return nil, errors.InvalidArgument(scheduler.EntityJobRun, "empty end date is given")
+	}
+	return &scheduler.JobRunsCriteria{
+		Name:      req.GetJobName(),
+		StartDate: req.GetStartDate().AsTime(),
+		EndDate:   req.GetEndDate().AsTime(),
+		Filter:    req.GetFilter(),
+	}, nil
+}
+
+func (h JobRunHandler) UploadToScheduler(ctx context.Context, req *pb.UploadToSchedulerRequest) (*pb.UploadToSchedulerResponse, error) {
+	projectName, err := tenant.ProjectNameFrom(req.GetProjectName())
+	if err != nil {
+		return nil, errors.GRPCErr(err, "unable to get projectName")
+	}
+	err = h.service.UploadToScheduler(ctx, projectName, req.GetNamespaceName())
+	if err != nil {
+		return nil, errors.GRPCErr(err, "unable to upload to scheduler for "+projectName.String())
+	}
+	return &pb.UploadToSchedulerResponse{
+		Status:       true,
+		ErrorMessage: "",
+	}, nil
+}
+
+func (h JobRunHandler) RegisterJobEvent(ctx context.Context, req *pb.RegisterJobEventRequest) (*pb.RegisterJobEventResponse, error) {
+	tnnt, err := tenant.NewTenant(req.GetProjectName(), req.GetNamespaceName())
```

Comment:
imo, this should be a early return call, the request should not be kept waiting white we persist the state or relay the notification. this will unnecessarily keep schedulers connections busy. 

@sbchaos @sravankorumilli